serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+        list_logit_bias_values = []
+
+        index_map = {}
+        idx_random = -1
+        idx_greedy = -1
+        batch_size = len(sampling_params)
+        # index 0 is for non-logprob requests
+        has_logprob = False
+        logprob_index_map = {}
+        list_mask_top_logprob = np.full(
+            ((LOGPROB_TOP_K_MAX) + 1, batch_size), False, dtype=bool
+        )
+        logit_bias_maxlen = 0
+        idxs_logprob = [-1] * ((LOGPROB_TOP_K_MAX) + 1)
+        for batch_idx, param in enumerate(sampling_params):
+            # Prepare temperature
+            # NOTE: Zero temperature means deterministic sampling
+            # (i.e., greedy sampling or beam search).
+            # Set the temperature to 1 to avoid division by zero.
+            list_temperatures.append(
+                param.temperature if param.temperature >= SAMPLING_EPS else 1.0
+            )
+
+            if param.sampling_type == SamplingType.RANDOM:
+                list_mask_random.append(True)
+                idx_random += 1
+                index_map[batch_idx] = (SamplingType.RANDOM, idx_random)
+                list_top_ps.append(param.top_p)
+                list_top_ks.append(param.top_k if param.top_k != -1 else vocab_size)
+                do_top_p |= list_top_ps[-1] < 1.0 - SAMPLING_EPS
+                do_top_k |= list_top_ks[-1] != vocab_size
+            else:
+                list_mask_random.append(False)
+                idx_greedy += 1
+                index_map[batch_idx] = (SamplingType.GREEDY, idx_greedy)
+
+            # param.top_logprobs is zero if logprob is not used
+            list_mask_top_logprob[param.top_logprobs] = True
+            has_logprob |= param.logprobs
+            idxs_logprob[param.top_logprobs] += 1
+            logprob_index_map[batch_idx] = (
+                param.logprobs,
+                param.top_logprobs,
+                idxs_logprob[param.top_logprobs],
+            )
+
+            apply_penalty |= (
+                abs(param.presence_penalty) >= SAMPLING_EPS
+                or abs(param.frequency_penalty) >= SAMPLING_EPS
+                or abs(param.repetition_penalty - 1.0) >= SAMPLING_EPS
+            )
+            list_frequency_penalties.append(param.frequency_penalty)
+            list_presence_penalties.append(param.presence_penalty)
+            list_repetition_penalties.append(param.repetition_penalty)
+
+            if param.logit_bias_index:
+                assert param.logit_bias_value
+                apply_bias |= True
+                logit_bias_maxlen = max(logit_bias_maxlen, len(param.logit_bias_index))
+                list_logit_bias_indices.append(param.logit_bias_index)
+                list_logit_bias_values.append(param.logit_bias_value)
+            else:
+                list_logit_bias_indices.append([])
+
+        num_random_samples = idx_random + 1
+        num_greedy_samples = idx_greedy + 1
+        assert num_random_samples + num_greedy_samples == batch_size
```

Comment:
Ideally change to an Exception type we can't catch for logging purposes.

Reply:
This assertion was for dev so I removed this. This is guaranteed to be passed.